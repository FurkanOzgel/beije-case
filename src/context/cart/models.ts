export interface Product {
    name: string;
    quantity: number;
    categoryID: number;
    oldQuantity: number;
}
export interface Cart {
    products: Product[];
}