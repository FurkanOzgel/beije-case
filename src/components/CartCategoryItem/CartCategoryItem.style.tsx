import { StyleSheet } from "react-native";

export default StyleSheet.create({
    categoryInfo: {
        flexDirection: "row",
        justifyContent: "space-between",
        alignItems: "center",
        padding: 24
    },
    productCategory: {
        fontSize: 18,
        fontWeight: '500',
        color: "rgba(0, 0, 0, 0.9)"
    },
    qunatityInfo: {
        fontSize: 14,
        color: "rgba(0, 0, 0, 0.6)",
        marginRight: 12
        
    }
});