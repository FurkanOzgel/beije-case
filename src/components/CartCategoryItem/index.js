export {default} from './CartCategoryItem.tsx';