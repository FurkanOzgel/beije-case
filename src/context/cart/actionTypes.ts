import {Product} from './models';
export const ADD_PRODUCT = 'ADD_PRODUCT';
export const DELETE_PRODUCT = 'DELETE_PRODUCT';
export const DELETE_CATEGORY = 'DELETE_CATEGORY';
export const UPDATE_OLD_QUANTITY = 'UPDATE_OLD_QUANTITY';


interface AddProductAction {
    type: typeof ADD_PRODUCT;
    payload: Product;
}

interface DeleteProductAction {
    type: typeof DELETE_PRODUCT;
    payload: Product;
}


interface DeleteCategoryAction {
    type: typeof DELETE_CATEGORY;
    payload: number;
}

interface UpdateOldQuantityAction {
    type: typeof UPDATE_OLD_QUANTITY;
    payload: Product;
}

export type CartActionTypes = AddProductAction | DeleteProductAction | DeleteCategoryAction | UpdateOldQuantityAction;