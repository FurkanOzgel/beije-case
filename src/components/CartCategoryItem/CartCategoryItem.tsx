import React from "react";
import { View, Text } from "react-native";

import { generateProductInfo } from "../../utils/helpers";
import { Cart } from "../../context/cart/models";

import icons from "../../assets/svg";
import styles from "./CartCategoryItem.style";

import IconButton from "../IconButton";

interface CartCategoryItemProps {
    cart: Cart;
    categoryID: number;
    onDeleteCategory: (categoryID: number) => void;
}

function CartCategoryItem(props: CartCategoryItemProps): React.JSX.Element {
    const { cart, categoryID, onDeleteCategory } = props;
    
    return(
        <View style={[styles.productInfo, !generateProductInfo(cart, categoryID) ? {display:'none'}:{}]}>
            <View>
                <Text style={styles.productCategory}>Ped Paketleri</Text>
                <Text style={styles.qunatityInfo}>{generateProductInfo(cart, categoryID)}</Text>
            </View>
            <IconButton svg={icons.bin} onPress={() => onDeleteCategory(categoryID)} />
        </View>
    )
}

export default CartCategoryItem;


