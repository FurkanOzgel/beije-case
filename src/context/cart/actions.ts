import {Product} from './models';
import {CartActionTypes, ADD_PRODUCT, DELETE_PRODUCT, DELETE_CATEGORY, UPDATE_OLD_QUANTITY} from './actionTypes';

export function addProduct(newProduct: Product): CartActionTypes {
    return {
        type: ADD_PRODUCT,
        payload: newProduct
    };
}

export function deleteProduct(product: Product): CartActionTypes {
    return {
        type: DELETE_PRODUCT,
        payload: product
    };
}

export function deleteCategory(categoryID: number): CartActionTypes {
    return {
        type: DELETE_CATEGORY,
        payload: categoryID
    };
}

export function updateOldQuantity(product: Product): CartActionTypes {
    return {
        type: UPDATE_OLD_QUANTITY,
        payload: product
    };
}