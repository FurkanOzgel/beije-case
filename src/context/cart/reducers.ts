import { Cart } from './models';
import {CartActionTypes, ADD_PRODUCT, DELETE_PRODUCT, DELETE_CATEGORY, UPDATE_OLD_QUANTITY} from './actionTypes';

const initialState: Cart = {
    products: []
};

export function cartReducer(
    state = initialState,
    action: CartActionTypes,
): Cart {
    switch (action.type) {
        case ADD_PRODUCT:
            const productIndex = state.products.findIndex(product => product.name === action.payload.name);

            if (productIndex !== -1) {
                const updatedProducts = state.products.map((product, index) => {
                    if (index === productIndex) {
                        return { ...product, quantity: product.quantity + action.payload.quantity };
                    }
                    return product;
                });
                return {
                    ...state,
                    products: updatedProducts
                };
            } else {
                return {
                    ...state,
                    products: [...state.products, { ...action.payload }]
                };
            }
        case DELETE_PRODUCT:
            const productIndexToDelete = state.products.findIndex(product => product.name === action.payload.name);
            if (productIndexToDelete !== -1) {
                if (state.products[productIndexToDelete].quantity > action.payload.quantity) {
                    const updatedProducts = state.products.map((product, index) => {
                        if (index === productIndexToDelete) {
                            return { ...product, quantity: product.quantity - action.payload.quantity };
                        }
                        return product;
                    });
                    return {
                        ...state,
                        products: updatedProducts
                    };
                } else {
                    return {
                        ...state,
                        products: state.products.filter((_, index) => index !== productIndexToDelete)
                    };
                }
            }
        case DELETE_CATEGORY:
            return {
                ...state,
                products: state.products.filter(product => product.categoryID !== action.payload)
            };
        case UPDATE_OLD_QUANTITY:
            const productToUpdateIndex = state.products.findIndex(product => product.name === action.payload.name);
            if (productToUpdateIndex !== -1) {
                const updatedProducts = state.products.map((product, index) => {
                    if (index === productToUpdateIndex) {
                        return { ...product, oldQuantity: action.payload.oldQuantity };
                    }
                    return product;
                });
                return {
                    ...state,
                    products: updatedProducts
                };
            }
        default:
        return state;
  }
}