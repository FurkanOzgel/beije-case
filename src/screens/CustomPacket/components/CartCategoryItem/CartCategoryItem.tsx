import React from "react";
import { View, Text } from "react-native";

import { generateProductInfo, getCategoryNameById } from "../../../../utils/helpers";
import { Cart } from "../../../../context/cart/models";

import icons from "../../../../assets/svg";
import styles from "./CartCategoryItem.style";

import IconButton from "../../../../components/IconButton";

interface CartCategoryItemProps {
    cart: Cart;
    categoryID: number;
    onDeleteCategory: (categoryID: number) => void;
}

function CartCategoryItem(props: CartCategoryItemProps): React.JSX.Element {
    const { cart, categoryID, onDeleteCategory } = props;
    
    return(
        <View style={[styles.categoryInfo, !generateProductInfo(cart, categoryID) ? {display:'none'}:{}]}>
            <View>
                <Text testID="category-title" style={styles.productCategory}>{getCategoryNameById(categoryID)}</Text>
                <Text style={styles.qunatityInfo}>{generateProductInfo(cart, categoryID)}</Text>
            </View>
            <IconButton svg={icons.bin} onPress={() => onDeleteCategory(categoryID)} />
        </View>
    )
}

export default CartCategoryItem;