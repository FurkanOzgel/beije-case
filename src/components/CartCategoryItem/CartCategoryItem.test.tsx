import { render, fireEvent, screen } from '@testing-library/react-native';
import CartCategoryItem from './CartCategoryItem';
import {Cart} from '../../context/cart/models';

const cart: Cart = {
    "products": [
        {
        "name": "product1",
        "quantity": 10,
        "categoryID": 2,
        "oldQuantity": 0
        }
    ]
}

test('should render correctly', () => {
    const container = render(<CartCategoryItem cart={cart} categoryID={2} onDeleteCategory={() => {}}/>);
    expect(container).toMatchSnapshot();
});
