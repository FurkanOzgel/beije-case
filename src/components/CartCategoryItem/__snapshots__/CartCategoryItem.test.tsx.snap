// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 24,
      },
      {},
    ]
  }
>
  <View>
    <Text
      style={
        {
          "color": "rgba(0, 0, 0, 0.9)",
          "fontSize": 18,
          "fontWeight": "500",
        }
      }
    >
      Günlük Ped Paketleri
    </Text>
    <Text
      style={
        {
          "color": "rgba(0, 0, 0, 0.6)",
          "fontSize": 14,
          "marginRight": 12,
        }
      }
    >
      10 product1
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="icon-button"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={24}
      bbWidth={24}
      fill="none"
      focusable={false}
      height={24}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 24,
            "width": 24,
          },
        ]
      }
      testID="svg"
      vbHeight={24}
      vbWidth={24}
      width={24}
      xml="<svg xmlns:xlink="http://www.w3.org/1999/xlink"  viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg" width="24" height="24"><path d="M16 6v-.8c0-1.12 0-1.68-.218-2.108a2 2 0 0 0-.874-.874C14.48 2 13.92 2 12.8 2h-1.6c-1.12 0-1.68 0-2.108.218a2 2 0 0 0-.874.874C8 3.52 8 4.08 8 5.2V6m2 5.5v5m4-5v5M3 6h18m-2 0v11.2c0 1.68 0 2.52-.327 3.162a3 3 0 0 1-1.311 1.311C16.72 22 15.88 22 14.2 22H9.8c-1.68 0-2.52 0-3.162-.327a3 3 0 0 1-1.311-1.311C5 19.72 5 18.88 5 17.2V6" stroke="#343131" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"></path></svg>"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M16 6v-.8c0-1.12 0-1.68-.218-2.108a2 2 0 0 0-.874-.874C14.48 2 13.92 2 12.8 2h-1.6c-1.12 0-1.68 0-2.108.218a2 2 0 0 0-.874.874C8 3.52 8 4.08 8 5.2V6m2 5.5v5m4-5v5M3 6h18m-2 0v11.2c0 1.68 0 2.52-.327 3.162a3 3 0 0 1-1.311 1.311C16.72 22 15.88 22 14.2 22H9.8c-1.68 0-2.52 0-3.162-.327a3 3 0 0 1-1.311-1.311C5 19.72 5 18.88 5 17.2V6"
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          propList={
            [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "payload": 4281610545,
              "type": 0,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth={2}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;
